### DISCUSSION TITLE
Memory leak detected (in Kubernetes)

### DISCUSSION THREAD

**Discussion by balazsorban44** [vercel/next.js#26801] [Help] [answered]
### What version of Next.js are you using?

10.2.3

### What version of Node.js are you using?

14.6.1

### What browser are you using?

Chrome (probably irrelevant)

### What operating system are you using?

Ubuntu 18.04 (see description for more info)

### How are you deploying your application?

Azure Kubernetes Service (AKS)

### Describe the Bug

Recently we went live with our new Next.js site, where we have ~200-300 users/hour traffic at peak times (8.00 - 16.00), and almost none (probably less than 10) throughout the rest of the day.

Right after deployment, the K8S pod started consuming increasingly more and more memory. See the image below:

![](https://i.imgur.com/whJ1tHe.png)

The interruptions are points where the pod itself either decided to restart, or we tried to fix the problem by pushing changes (see below). The memory usage drops to an initially low value when restarting, but it starts creeping up again. It seems to happen out of our peak hours as well (although most obvious during higher traffic).

I searched through issues and PRs, here are some potentially relevant ones:
- #22644
- #20915
- #24282

We made sure to be on a version that included the fixes (10.2.3), and I replaced all our `Image` components with `img`. The problem did not disappear.

I can confirm that the memory leak comes from `next` because I can ssh into the pod and using [`pmap`](https://linux.die.net/man/1/pmap), I read that ~80-90% of the RAM is allocated to the process running `next`.

If anyone has any tips on how to debug this further, that would be very helpful!

### Expected Behavior

I expect the RAM usage not to grow continuously, especially when there is a very low number of users out of peak hours.

### To Reproduce

Using `AKSUbuntu-1804containerd-2021.06.02` which has the following content:
https://github.com/Azure/AKS/blob/2021-06-03/vhd-notes/aks-ubuntu/AKSUbuntu-1804/2021.06.02.txt

I would like to provide even more information, but our codebase is private. I could try to ask us to set up a session/meeting if someone would be willing to help.

**Comment by jamsinclair** [comment]
@balazsorban44 the best next step is to start capturing heap dumps/snapshots to see what is being retained in the memory and contributing to the large size.

You're using node.js v14 so you should be able to make use of the native API `v8.getHeapSnapshot()`, see https://dev.to/bengl/node-js-heap-dumps-in-2021-5akm. You'll also want to read up on other articles on how to analyze leaks from the snapshot data.

If you can't replicate the leak locally you could try:
1. Create a private prod-like deployment
2. Run load tests against it, similarly to your production traffic. I can recommend the tool [k6](https://k6.io/)
3. Periodically capture heap snapshots from your app code
4. Download the snapshots and analyse the objects in memory

The thing to diagnose is whether the leaking memory is related to your app code or a problem in Next.js itself. Best of luck! 🤞 (The next process is running your code, so it doesn't necessarily mean the problem is Next 😉 )

**Reply by balazsorban44** [reply]
Thank you for your tips! I think I managed to narrow it down to a single API endpoint.

I created a clean Next.js project with this single API route:

```js
import pino from "pino"
import pinoToSeq from "pino-seq"

const stream = pinoToSeq.createStream({
  apiKey: process.env.API_KEY,
  serverUrl: process.env.SERVER_URL,
})

const log = pino({ level: "debug" }, stream).child({
  BUILD_ID: process.env.BUILD_ID || "dev",
  BUILD_SOURCE_BRANCH: process.env.BUILD_SOURCE_BRANCH,
  BUILD_SOURCE_VERSION: process.env.BUILD_SOURCE_VERSION,
})

export default function test(_, res) {
  log.debug("Balazs tester")
  log.debug("Balazs tester")
  log.debug("Balazs tester")
  log.debug("Balazs tester")
  log.debug("Balazs tester")
  log.debug("Balazs tester")

  res.end()
}
```

We are using pino, and Seq, but I also tried Sentry, the memory increase has the same pattern:

![image](https://user-images.githubusercontent.com/18369201/124619242-c48dc800-de78-11eb-8151-e75456f614e4.png)

(We used `autocannon` to bombard the endpoint with a couple of thousand requests to build up the RAM usage)

After reading a ton about `stream.Writeable`, I think the problem comes from that. If I don't forward the logs to a third party (only logging them to the terminal), the memory usage does not increase, at least not the same speed?!:

![](https://i.imgur.com/U1Ni3Qi.png)

I tried to close/flush/cork/uncork/end the stream in many ways, but either it had no effect, or the performance dropped ten folds... I haven't used `stream` before, so I walk on thin ice.

**Reply by jamsinclair** [reply]
Great work I think you've found the cause of the leak!

How often do you log throughout your application code?

As noted in the Pino documentation:
> Due to Node's single-threaded event-loop, it's highly recommended that sending, alert triggering, reformatting and all forms of log processing is conducted in a separate process.
> - https://github.com/pinojs/pino#transports--log-processing 

If you're constantly streaming logs you may want to set up a sidecar container or another mechanism to handle the logs and stream to your logging server.

**Reply by balazsorban44** [reply]
Since we are running in Kubernetes, today we decided to skip logging to our third-party at all, and only send logs to the console. From there, we extract the logs with [Grafana Loki](https://grafana.com/oss/loki/) and present it in a Grafana Dashboard, in hope that the bottleneck might have been our third-party logging service not being able to handle the incoming requests, which would clog up our Next.js app, filling up the RAM. (This is how I understood `stream.Writable` at least. When the response is slow, it collects the requests in memory and waits until the receiving party is ready again.)

The reason being that when conducting the same stress test locally, the results showed a much less steep increase in RAM usage:

![](https://i.imgur.com/XAn9JK6.png)

I have now run this in production for a few hours, but the **MEMORY STILL LEAKS**. 😕

Meaning that the problem might not be the logging service after all...

Maybe because we sent so many requests, we artificially created an unlikely high load on our logging service, which caused it to respond slowly.

Even after turning down the log level at the production app and excluding any kind of transport, the RAM usage still increases constantly.

![](https://i.imgur.com/Ro48yaB.png)
The orange line where we turned down the logging level.
The green line is the newest pod, with only logging to the console.

I did keep `pino` to format the logs, so I don't know if your comment is still relevant? Or that would only matter if I used a Transport?

**Comment by balazsorban44** [comment]
I managed to get some memory dumps from our production pod! (freezing the whole site for 30-40sec after busy hours... phew. 😅)

Some screenshots:
![](https://i.imgur.com/pqzMR3N.png)

![](https://i.imgur.com/5S6X2Cb.png)

(Heapdumps were taken when the server has only run a few hours. In peak hours, the memory usage already neared the limit before it would crash the server. When I ran took the heapdump then, the server crashed.)

To me, it looks that the Next.js `IncrementalCache` is not getting emptied!
We use this version in prod: https://github.com/vercel/next.js/blob/v11.0.2-canary.3/packages/next/next-server/server/incremental-cache.ts

My assumption was that this will flush the generated content to the disk, but it looks like it is kept in memory indefinitely? @timneutkens could you explain how/when the `lru-cache` is emptied? And if it should but it doesn't in my case, what information could I provide to find the problem?

**Reply by jamsinclair** [reply]
The `IncrementalCache` uses an in-memory LRU cache (Least Recently Used). This means it'll grow to its max size, 50MB, and stay there indefinitely unless data is manually deleted. When at the limit, newer data will cause the lesser-used data to be deleted. The cache is never emptied by the next.js server (from my understanding). 

You can anticipate that part of your Next.js server memory usage could grow +50MB due to this cache. Your dump shows it at around 45MB, which seems legitimate. 

I noticed in one of your graphs you have quite a high number of active node.js handles. Do you have metrics for these from the production server and does it remain high? Maybe a clue 🔍 , but yeah leaks are tough to debug 😩 .

**Reply by balazsorban44** [reply]
If you mean the previous comment, that is where we bombarded one of our API endpoints (that we thought was the problem) with 1000 connections for 30 sec.

Turns out our tests were flawed. We did see an increase in memory usage because the underlying third party could not handle hundreds of thousands of requests in those 30 secs. Even after changing to `stdout`, in production, we can still experience the memory leakage so it was unrelated after all.

**Reply by balazsorban44** [reply]
So I checked, and the `IncrementalCache` does grow bigger than 50MB!

![](https://i.imgur.com/CMicabk.png)

**Comment by balazsorban44** [comment, marked as answer]
I can confirm that our production deployment has stabilized after setting the following (as per https://github.com/vercel/next.js/pull/27337) in `next.config.js`:
```js
experimental: {
 isrMemoryCacheSize: 0 
}
```

So the problem really seems to originate from Next.js itself.

My question is, what is the point of the ISR cache actually?

If a sufficient CDN sits in front of the Next.js origin server, any request reaching the origin should re-generate a new page when ISR revalidation is used. So the LRU cache is never really utilized for anything. Am I misunderstanding something?

**Reply by jamsinclair** [reply]
Thank you, interesting! We're still debugging a memory leak with our Production servers as well, will give this a try next week and confirm whether it makes a difference for us! 🤔

> If a sufficient CDN sits in front of the Next.js origin server, any request reaching the origin should re-generate a new page when ISR revalidation is used. So the LRU cache is never really utilized for anything. Am I misunderstanding something?

In that use case the Next.js LRU cache is a little redundant as you have another caching layer, the CDN. 

The advantage is when you're not using a CDN, which is the case in my situation, the server can quickly return a pre-rendered page from memory and save on time having to read and buffer from the file system.

**Reply by balazsorban44** [reply]
Interesting! I always assumed that Vercel uses a CDN in front of their Next.js applications! 😯

**Reply by balazsorban44** [reply]
Here is the code we used to generate heapdumps in production, which used to be very helpful in debugging:

```js
import v8 from "v8"
import fs from "fs"
import path from "path"

export default async function heapDump(req, res) {
  try {
    if (isAdmin(req)) { // Determine if the request is authorized to avoid malicious intent
      return res.status(403).end("Access denied.")
    }

    if (req.query.filename) {
      const { filename } = req.query
      const filePath = path.join(process.cwd(), filename)
      res.setHeader("Content-Type", "application/json")
      res.setHeader("Content-Disposition", `attachment; filename=${filename}`)
      const file = fs.readFileSync(filePath)
      return res.end(file)
    }

    const filename = v8.writeHeapSnapshot()

    res.setHeader("Content-Type", "text/html")
    res.end(
      `<h1>Heapdump generated. Download it from here: <a href="/api/heapdump?filename=${filename}">${filename}</a></h1>`
    )
  } catch (error) {
    console.error(error)
    res.end("An error occured.")
  }
}
```

Keep in mind that this will block the server for a while and increase the overall RAM usage after it has run. To download the generated file, just click the link on the resulting page. For us, the page actually timed out after 30sec, so we had to find the file in the K8S pod to get the generated filename and invoked the endpoint with the `filename` query param.

If you have multiple instances in production, it might take some trial/error to get the heapdump this way, or you can use SSH

**Reply by jamsinclair** [reply]
> Interesting! I always assumed that Vercel uses a CDN in front of their Next.js applications! 😯

I'm not at Vercel 😅 , just another unrelated company. I believe they do use the CDN & serverless approach for the Vercel platform.   

Thanks for the tips! We had a similar approach for generating our memory heap dumps. After some time investigating our memory leak we think it's unrelated to the Incremental Cache, we've tried both with it enabled and disabled. 

We're not 100% confident but it looks like our memory is growing in response to the Linux filesystem buffer cache. Over time, Next.js is writing more and more files to our Kubernetes pod and we see that reflected in the memory usage of the pod.  This would make sense, we're operating at scale and over 1GB of page data so far is being written to disk.
